File: backend/app/insight_types.py
"""Canonical user-facing insight types — the categories a PM picks to say which
findings they want as their Top Insights.

Single source of truth on the backend for:
  * the top-insights compose prompt (each composed finding is classified into
    one or more of these, so filtering matches the SAME vocabulary the user
    picked from — see synthesis/agent.py),
  * the per-user preference store and its validation,
  * the DB CHECK constraints on the stored preference (kept byte-identical in
    the migration that introduces them).

Mirrors the frontend list in web/app/lib/insight-types.ts. Adding, removing, or
renaming a type means changing BOTH sides AND the DB constraint(s).

THE TWO LISTS MUST BE THE SAME LENGTH. A type the backend can classify a
finding into, but the frontend does not render, is worse than no type at all:
the compose prompt promotes a finding under it and the browser then filters
that finding out of view, with nothing on either surface saying why. Between
2026-07-27 and 2026-08-05 exactly that held — the backend counted six, the
picker offered three — and the extra three were invisible state. Apurva's
ruling on 2026-08-05: a backend insight type is either wired through to the
web or it does not exist.

History: the original 6 onboarding chips merged with 3 client-requested report
types (2026-07-23; all three were duplicates, so the merged set stayed six),
then narrowed to the THREE the picker actually offers (2026-08-05). The three
dropped — user_feedback, reliability_signals, wins — had no skill configured to
produce them and no chip to select them. Two slugs were renamed in the 07-23
merge as their meaning broadened:
  drive_metric        -> build_priorities
  emerging_complaints -> user_feedback (since dropped)
See the accompanying migration for the data remap.

The DB CHECK constraints deliberately still accept the wider vocabulary. A
constraint that is a strict superset of the code is safe — nothing can write a
slug this module does not know — whereas narrowing it needs a migration and a
data remap for rows that already hold a dropped slug. `clean_insight_types`
drops those on read, which degrades to "surface everything", the same default
as no preference at all.
"""
from __future__ import annotations

#: slug -> (label, one-line description). The description is fed to the compose
#: prompt verbatim so the model classifies each finding into the exact same
#: categories the user selects from, making the per-user filter precise rather
#: than a fuzzy mapping off the internal 7-way skill taxonomy.
#: Order is the picker's chip order, so the prompt, the chips and the settings
#: rows all read the same way round.
INSIGHT_TYPES: "dict[str, tuple[str, str]]" = {
    "top_problems": (
        "Top user problems & opportunities",
        "The most pressing user/product problems and the biggest opportunities "
        "surfaced across all signals.",
    ),
    "competitor_moves": (
        "Competitor & market moves",
        "Competitive and market developments the team should react to "
        "(launches, pricing, positioning, category shifts).",
    ),
    "build_priorities": (
        "Most important to build",
        "The highest-priority things to build next, synthesizing every signal "
        "(metric movement, user demand, revenue, strategy).",
    ),
}

#: The slugs, in canonical display order. Use this everywhere a fixed set is
#: needed (schema enum, constraint list, validation).
INSIGHT_TYPE_SLUGS: "tuple[str, ...]" = tuple(INSIGHT_TYPES.keys())

#: slug -> (badge text, accent hex) for the CARD PILL the reader sees.
#:
#: Until 2026-08-05 the pill showed the top-insights skill's own 8-way taxonomy
#: (Reliability, Growth, Demand, Retention, Competitive, Engagement, Compliance,
#: Momentum) — a vocabulary the preference picker does not contain, so a reader
#: who asked for "Reliability & incident signals" had no way to look at a card
#: and tell whether their selection had been honoured. GROWTH in particular is
#: not a preference slug at all. The pill now names the finding's OWN
#: `insight_types`, i.e. the exact vocabulary the picker offers.
#:
#: The badge text is a short form of the picker's chip label (the chip wording
#: in full is too long for an 11px uppercase pill); it lives here beside the
#: prompt label so the two can never drift. Accents are reused verbatim from the
#: skill taxonomy's existing palette — no new colours are introduced, each slug
#: simply claims the hex whose meaning already matched it.
INSIGHT_TYPE_BADGES: "dict[str, tuple[str, str]]" = {
    "top_problems":     ("Top problem",      "#b23b52"),  # rose
    "competitor_moves": ("Competitor moves", "#b07a2e"),  # ochre
    "build_priorities": ("What to build",    "#1a8a52"),  # green
}


def is_valid_insight_type(slug: str) -> bool:
    return slug in INSIGHT_TYPES


def display_insight_type(
    insight_types: object, selected: "list[str] | None" = None,
) -> "str | None":
    """Which of a finding's insight types to show on its card.

    A finding carries one or two, in the model's own order — the first is its
    PRIMARY classification. We walk the finding's types in that order and take
    the first one the reader selected, so a card whose primary type was asked
    for keeps it, and a card whose primary was NOT asked for surfaces the
    secondary type that was. Walking the SELECTION order instead would let a
    reader's first chip override every card's primary and collapse distinct
    findings to the same label. With no selection, the primary.

    Returns None when the finding carries no known type, which
    is the legacy case: briefs composed before the classifier existed have no
    `insight_types` at all, and the caller must keep its old skill-taxonomy label
    rather than invent one (the 8 skill types do not map cleanly onto the 3
    preference slugs — retention, demand, engagement and compliance each have no
    faithful counterpart).
    """
    types = clean_insight_types(insight_types)
    if not types:
        return None
    wanted = set(selected or ())
    for slug in types:
        if slug in wanted:
            return slug
    return types[0]


def clean_insight_types(values: object) -> "list[str]":
    """Filter an arbitrary input down to known slugs, order-preserving and
    de-duplicated. Returns [] for anything unusable — the readers treat an empty
    selection as "surface everything", so a junk value degrades to the default
    rather than raising."""
    if not isinstance(values, (list, tuple)):
        return []
    out: list[str] = []
    for v in values:
        if isinstance(v, str) and v in INSIGHT_TYPES and v not in out:
            out.append(v)
    return out


def order_pool_for_types(
    pool: "list[dict]", selected: "list[str]",
) -> "tuple[list[dict], int]":
    """Stable partition of a composed pool by the reader's insight types.

    Findings whose `insight_types` intersect `selected` lead, in their existing
    (best-first) order; everything else follows, also in its existing order.
    Returns the reordered pool plus how many findings matched.

    This is the DETERMINISTIC half of the preference contract. The compose
    prompt already carries the selection as a ranking nudge
    (synthesis/reader_prefs), but a nudge can't guarantee the LEAD finding
    matches — measured across the live briefs, a preferred finding existed but
    sat below rank 1 in most of them. Reordering here makes `insights[0]` — the
    canonical top insight the weekly email, the Slack post, PRD warming and the
    KG ledger all key off — a preferred finding whenever the pool holds one.

    Honours SKILL.md step 4b literally: preferences REORDER, they never
    exclude. Nothing is dropped, and no selection (or no match) returns the
    pool unchanged so the model's own ranking stands. Byte-for-byte the same
    semantics as the frontend's orderPoolForTypes in web/app/lib/
    brief-v2-adapter.ts, so the browser's partition of an already-partitioned
    pool is the identity and the two surfaces cannot disagree.
    """
    wanted = {s for s in selected if s in INSIGHT_TYPES}
    # Nothing selected, or EVERYTHING selected — both mean "no preference", so
    # the model's own ranking stands. The all-types case is not redundant: a
    # legacy finding carries no `insight_types`, so it intersects no selection
    # and would be demoted below every classified finding by a selection that
    # was meant to express no preference at all. The pickers now resolve a
    # cleared selection to the full set rather than to [], so this is the shape
    # that actually arrives. Mirrors coversEveryInsightType in the frontend's
    # lib/insight-types.
    if not wanted or wanted >= set(INSIGHT_TYPES):
        return pool, 0
    matching: list[dict] = []
    rest: list[dict] = []
    for ins in pool:
        types = ins.get("insight_types")
        if isinstance(types, (list, tuple)) and wanted.intersection(types):
            matching.append(ins)
        else:
            rest.append(ins)
    return matching + rest, len(matching)


def prompt_block() -> str:
    """The TYPES reference block injected into the compose prompt, so the model
    classifies each finding into these exact categories."""
    lines = ["INSIGHT TYPES — classify every finding into one or two of these:"]
    for slug, (label, desc) in INSIGHT_TYPES.items():
        lines.append(f"  - {slug} ({label}): {desc}")
    return "\n".join(lines)

File: web/app/lib/insight-types.ts
// Canonical user-facing insight types — the categories a PM picks to say which
// findings they want as their Top Insights. Single source of truth on the
// frontend for the onboarding chips, the settings pane, and the inline picker
// on the Top Insights tab.
//
// Mirrors backend/app/insight_types.py on the SLUGS — those are the contract
// (stored preference, DB check constraint, the vocabulary the compose prompt
// classifies into), so adding or removing a type means changing BOTH sides AND
// the DB constraint(s). The `label` is product wording for the chips and is
// deliberately allowed to differ from the backend's: the backend label is
// prompt text the classifier reads, and rewording it there would shift how
// findings get categorised.
//
// THERE IS ONE LIST. Until 2026-08-05 there were two — a canonical six and a
// selectable three — and the gap between them was invisible state: the backend
// classified findings into all six and promoted them, and this file then
// filtered the extra three out of view, with nothing on either surface saying
// why a promoted finding never appeared. Apurva's ruling on 2026-08-05: a
// backend insight type is either wired through to the web or it does not
// exist. Do not reintroduce a "selectable subset" — put the type in this list
// when it goes live, and not before.
//
// History: the original 6 onboarding chips merged with 3 client-requested
// report types (2026-07-23; all three were duplicates, so the merged set
// stayed six), then narrowed to the three the picker actually offers
// (2026-08-05). The three dropped — user_feedback, reliability_signals, wins —
// had no skill configured to produce them. Two slugs were renamed in the 07-23
// merge as their meaning broadened: drive_metric -> build_priorities,
// emerging_complaints -> user_feedback (since dropped).
//
// Stored rows may still hold a dropped slug: the DB constraint deliberately
// stays wider than the code, and `cleanInsightTypes` drops unknown slugs on
// read. A selection that empties out means "surface everything" — the same
// default as no preference at all.

export type InsightTypeSlug =
  | "top_problems"
  | "competitor_moves"
  | "build_priorities"

export interface InsightType {
  value: InsightTypeSlug
  label: string
  /** Short helper copy for the chip's tooltip / settings row description. */
  description: string
}

// Order here is display order everywhere the chips render.
export const INSIGHT_TYPES: InsightType[] = [
  {
    value: "top_problems",
    label: "Top Customer Problem",
    description:
      "The most pressing user and product problems, and the biggest opportunities across your signals.",
  },
  {
    value: "competitor_moves",
    label: "Competitor & market moves",
    description: "Competitive and market developments worth reacting to.",
  },
  {
    value: "build_priorities",
    label: "What to build next",
    description:
      "The highest-priority things to build next, weighing every signal — metrics, demand, revenue, strategy.",
  },
]

export const INSIGHT_TYPE_SLUGS: InsightTypeSlug[] = INSIGHT_TYPES.map((t) => t.value)

const INSIGHT_TYPE_SET = new Set<string>(INSIGHT_TYPE_SLUGS)
const LABEL_BY_SLUG: Record<string, string> = Object.fromEntries(
  INSIGHT_TYPES.map((t) => [t.value, t.label]),
)

// The CARD PILL — badge text + accent for a finding, in the reader's own
// vocabulary. Mirrors backend/app/insight_types.py INSIGHT_TYPE_BADGES.
//
// Until 2026-08-05 the pill showed the top-insights skill's separate 8-way
// taxonomy (Reliability, Growth, Demand, Retention, Competitive, Engagement,
// Compliance, Momentum). That vocabulary is not what the picker offers —
// "Growth" is not a preference type at all — so a reader who asked for
// "Reliability & incident signals" could not look at a card and tell whether
// their selection had been honoured. The pill now names the finding's own
// `insight_types`.
//
// `badge` is a short form of `label` above: the chip wording in full does not
// fit an 11px uppercase pill. It lives beside the chip label so the two cannot
// drift — swap these strings for `label` if the full wording is wanted.
// Accents are the skill taxonomy's existing hexes, reassigned by meaning; no
// new colours are introduced.
export const INSIGHT_TYPE_BADGES: Record<InsightTypeSlug, { badge: string; accent: string }> = {
  top_problems: { badge: "Top problem", accent: "#b23b52" }, // rose
  competitor_moves: { badge: "Competitor moves", accent: "#b07a2e" }, // ochre
  build_priorities: { badge: "What to build", accent: "#1a8a52" }, // green
}

/** Which of a finding's insight types to show on its card.
 *
 *  A finding carries one or two, in the model's own order — the first is its
 *  PRIMARY classification. We walk the finding's types in that order and take
 *  the first one the reader selected, so a card whose primary type was asked
 *  for keeps it, and a card whose primary was NOT asked for surfaces the
 *  secondary type that was. Walking the SELECTION order instead would let a
 *  reader's first chip override every card's primary and collapse distinct
 *  findings to the same label. With no selection, the primary.
 *
 *  Returns null for a legacy finding with no `insight_types` — the caller must
 *  keep its old skill-taxonomy label rather than invent one, since the 8 skill
 *  types have no faithful counterpart for retention, demand, engagement or
 *  compliance. Mirrors backend/app/insight_types.py display_insight_type. */
export function displayInsightType(
  insightTypes: unknown,
  selectedTypes: string[] = [],
): InsightTypeSlug | null {
  const types = cleanInsightTypes(insightTypes)
  if (types.length === 0) return null
  const wanted = new Set(selectedTypes)
  for (const slug of types) {
    if (wanted.has(slug)) return slug
  }
  return types[0]
}

export function isInsightTypeSlug(v: unknown): v is InsightTypeSlug {
  return typeof v === "string" && INSIGHT_TYPE_SET.has(v)
}

/** Keep only known slugs, order-preserving + de-duplicated. Unknown/garbage
 *  input degrades to [] — an empty selection means "surface everything", the
 *  same default the readers use when no preference is stored. */
export function cleanInsightTypes(values: unknown): InsightTypeSlug[] {
  if (!Array.isArray(values)) return []
  const out: InsightTypeSlug[] = []
  for (const v of values) {
    if (isInsightTypeSlug(v) && !out.includes(v)) out.push(v)
  }
  return out
}

/** What a workspace starts with before anyone has picked. ONE constant, read by
 *  both screens that write `brief_insight_types` — onboarding step 09 and
 *  Settings → Comms & Brief.
 *
 *  They used to disagree: onboarding seeded these two, Settings seeded []. Same
 *  key, two different starting states, and since [] means "surface everything"
 *  they were not just different but OPPOSITE — one screen opened filtered, the
 *  other unfiltered, for a workspace in the identical state. */
export const DEFAULT_INSIGHT_TYPES: InsightTypeSlug[] = [
  "top_problems",
  "build_priorities",
]

/** Does this selection already cover every type we know about?
 *
 *  Load-bearing, not cosmetic. "All types selected" and "nothing selected" MUST
 *  behave identically, and they do not do so by accident: a legacy finding
 *  carries no `insight_types` at all, so it intersects no selection and would be
 *  filtered OUT by an explicit all-types list while an empty list keeps it. 30
 *  of the 90 findings rendered across the live briefs are in that state, so
 *  treating the two as the same thing is what makes "cleared = use all" safe. */
export function coversEveryInsightType(selection: readonly string[]): boolean {
  const picked = new Set(cleanInsightTypes(selection))
  return INSIGHT_TYPE_SLUGS.every((slug) => picked.has(slug))
}

/** Resolve the stored value into what the picker shows and the readers use.
 *
 *  THE SELECTION IS NEVER EMPTY. Three cases:
 *    - key absent / not an array  => nobody has chosen yet => DEFAULT_INSIGHT_TYPES
 *    - cleans to nothing          => cleared => EVERY type ("use all")
 *    - otherwise                  => the cleaned selection
 *
 *  Clearing every chip is a request to stop filtering, not a request for an
 *  empty brief, so it resolves to the full set rather than to `[]`. That keeps
 *  the state legible — the picker always shows what you will get, and a stored
 *  row always says it outright instead of encoding "everything" as absence.
 *
 *  Legacy rows holding `[]` (written before this rule) resolve the same way, so
 *  they need no backfill: they already meant "everything". */
export function resolveInsightTypes(values: unknown): InsightTypeSlug[] {
  if (!Array.isArray(values)) return [...DEFAULT_INSIGHT_TYPES]
  const cleaned = cleanInsightTypes(values)
  return cleaned.length ? cleaned : [...INSIGHT_TYPE_SLUGS]
}

export function insightTypeLabel(slug: string): string {
  return LABEL_BY_SLUG[slug] ?? slug
}
